### PR TITLE
Refactor Execute() of Allocate Method into Multiple Subfunctions for Enhanced Readability and Maintainability

### PR DESCRIPTION
**Pull request by daniel-hutao**
## Summary

This pull request refactors a large core function into multiple subfunctions. The primary goal is to enhance the code's readability and maintainability.

This work is not yet fully completed, but considering the significant amount of changes already made, I hope to merge the current changes first to facilitate an easier review process. I plan to continue further optimization subsequently.

This PR is divided into two commits, with each commit presenting relatively clear changes. I would kindly ask the reviewers to consider reviewing the changes commit by commit, as this approach may be more manageable.

### CODE REVIEW

File: pkg/scheduler/actions/allocate/allocate.go [hunk 1]
```diff
@@ -57,6 +57,15 @@ func (alloc *Action) Execute(ssn *framework.Session) {
 	// jobsMap is used to find job with the highest priority in given queue.
 	jobsMap := map[api.QueueID]*util.PriorityQueue{}
 
+	alloc.prepareQueuesAndJobs(queues, jobsMap, ssn)
```
Comment (Monokaix) [inline pkg/scheduler/actions/allocate/allocate.go:60]: We can rename it something like `pickUpQueuesAndJobs`

File: pkg/scheduler/actions/allocate/allocate.go [hunk 2]
```diff
@@ -164,123 +156,146 @@ func (alloc *Action) Execute(ssn *framework.Session) {
...
-			if !ssn.Allocatable(queue, task) {
-				klog.V(3).Infof("Queue <%s> is overused when considering task <%s>, ignore it.", queue.Name, task.Name)
-				continue
-			}
+func (alloc *Action) processTasks(tasks *util.PriorityQueue, job *api.JobInfo, jobs *util.PriorityQueue, queue *api.QueueInfo, ssn *framework.Session, allNodes []*api.NodeInfo) {
+	stmt := framework.NewStatement(ssn)
+	ph := util.NewPredicateHelper()
 
-			klog.V(3).Infof("There are <%d> nodes for Job <%v/%v>", len(ssn.Nodes), job.Namespace, job.Name)
+	predicateFn := func(task *api.TaskInfo, node *api.NodeInfo) ([]*api.Status, error) {
```
Comment (Monokaix) [inline pkg/scheduler/actions/allocate/allocate.go:167]: Before this modification, predicateFn just declared once, it's declared repeatedly now, we should keep declare it once.

File: pkg/scheduler/actions/allocate/allocate.go (hunk 1 above)
Comment (daniel-hutao) [reply pkg/scheduler/actions/allocate/allocate.go:60]: good idea

File: pkg/scheduler/actions/allocate/allocate.go (hunk 2 above)
Comment (daniel-hutao) [reply pkg/scheduler/actions/allocate/allocate.go:167]: yeah

**Comment by daniel-hutao** [comment]
@Monokaix Please take a look again.

File: pkg/scheduler/actions/allocate/allocate.go [hunk 3]
```diff
@@ -90,29 +100,13 @@ func (alloc *Action) Execute(ssn *framework.Session) {
 		klog.V(4).Infof("Added Job <%s/%s> into Queue <%s>", job.Namespace, job.Name, job.Queue)
 		jobsMap[job.Queue].Push(job)
 	}
+}
 
-	klog.V(3).Infof("Try to allocate resource to %d Queues", len(jobsMap))
-
+func (alloc *Action) allocateResources(queues *util.PriorityQueue, jobsMap map[api.QueueID]*util.PriorityQueue) {
```
Comment (Monokaix) [inline pkg/scheduler/actions/allocate/allocate.go:105]: This method does two things, firstly pickUpTasks and allocate task secondly, so rename this method name to pickUpTasks  and rename processTasks to allocateResourcesForTasks, and both make them as a dependent method seperately is better, just be consistent with `alloc.pickUpQueuesAndJobs`.

Comment (daniel-hutao) [reply pkg/scheduler/actions/allocate/allocate.go:105]: The main logic of the `allocateResources` method is in a **for loop**, and a longer part of the logic within the loop has been separated into `processTasks`. I attempted to decouple these two methods and have them both called by the `Execute` method, but this failed.

Moving the for loop logic into `Execute` essentially means removing the `allocateResources` method, which would significantly complicate `Execute`. Therefore, I think it's not suitable to rename `allocateResources` to `pickUpTasks`.

As for renaming `processTasks`, I am open to it. If you think a change would be better, I can add another commit to complete this renaming.

Comment (Monokaix) [reply pkg/scheduler/actions/allocate/allocate.go:105]: Got that.

**Review by Thor-wl — APPROVED** [review]

**Pull request merged**